Source/controls/keymapper.cpp

```diff
@@ -0,0 +1,126 @@
+#include "keymapper.hpp"
+
+#include <cassert>
+#include <fmt/format.h>
+
+#include <SDL.h>
+
+#ifdef USE_SDL1
+#include "utils/sdl2_to_1_2_backports.h"
+#endif
+
+#include "../miniwin/miniwin.h"
+#include "utils/log.hpp"
+
+namespace devilution {
+
+Keymapper::Keymapper(SetConfigKeyFunction setKeyFunction, GetConfigKeyFunction getKeyFunction)
+    : setKeyFunction(setKeyFunction)
+    , getKeyFunction(getKeyFunction)
+{
+	keyIDToKeyName.reserve('Z' - 'A' + '9' - '0' + 12);
+	for (char c = 'A'; c <= 'Z'; ++c) {
+		keyIDToKeyName.emplace(c, std::string(1, c));
+	}
+	for (char c = '0'; c <= '9'; ++c) {
+		keyIDToKeyName.emplace(c, std::string(1, c));
+	}
+	for (int i = 0; i < 12; ++i) {
+		keyIDToKeyName.emplace(DVL_VK_F1 + i, fmt::format("F{}", i + 1));
+	}
+
+	keyNameToKeyID.reserve(keyIDToKeyName.size());
+	for (const auto &kv : keyIDToKeyName) {
+		keyNameToKeyID.emplace(kv.second, kv.first);
+	}
+}
+
+Keymapper::ActionIndex Keymapper::addAction(const Action &action)
+{
+	actions.emplace_back(action);
+
+	return actions.size() - 1;
+}
+
+void Keymapper::keyPressed(int key, bool isPlayerDead) const
+{
+	auto it = keyIDToAction.find(key);
+	if (it == keyIDToAction.end())
+		return; // Ignore unmapped keys.
+
+	auto &action = it->second;
+
+	// Some actions cannot be triggered while the player is dead.
+	if (isPlayerDead && action.get().ifDead == Action::IfDead::Ignore)
+		return;
+
+	action();
+}
+
+std::string Keymapper::keyNameForAction(ActionIndex actionIndex) const
+{
+	assert(actionIndex < actions.size());
+	auto key = actions[actionIndex].key;
+	auto it = keyIDToKeyName.find(key);
+	assert(it != keyIDToKeyName.end());
+	return it->second;
+}
+
+void Keymapper::save() const
+{
+	// Use the action vector to go though the actions to keep the same order.
+	for (const auto &action : actions) {
+		if (action.key == DVL_VK_INVALID) {
+			// Just add an empty config entry if the action is unbound.
+			setKeyFunction(action.name, "");
+			continue;
+		}
+
+		auto keyNameIt = keyIDToKeyName.find(action.key);
+		if (keyNameIt == keyIDToKeyName.end()) {
+			Log("Keymapper: no name found for key '{}'", action.key);
+			continue;
+		}
+		setKeyFunction(action.name, keyNameIt->second);
+	}
+}
+
+void Keymapper::load()
+{
+	keyIDToAction.clear();
+
+	for (auto &action : actions) {
+		auto key = getActionKey(action);
+		action.key = key;
+		if (key == DVL_VK_INVALID) {
+			// Skip if the action has no key bound to it.
+			continue;
+		}
+		// Store the key in action.key and in the map so we can save() the actions while keeping the same order as they have been added.
+		keyIDToAction.emplace(key, action);
+	}
+}
+
+int Keymapper::getActionKey(const Keymapper::Action &action)

```

Comment:
Does this still allow you to unbind actions that you don't want to use, despite them having a default key?

Reply:
Yes, the key that is set if it's unbound is `DVL_VK_INVALID`. You can add `MyAction=` in the config file to unbind an action.